### ISSUE TITLE
Syntax Highlighting

### ISSUE THREAD

**Issue by ashn-dot-dev** [teal-language/tl#60]
It would be nice to have Syntax highlighting for tl in various editors. I suspect turning on Lua syntax highlighting will cover most bases (works okay in Vim), but specialized keyword highlighting would be a welcomed quality-of-life improvement.

Not sure if this should be included as part of this repository or maintained in other repo(s).

Also not sure if syntax highlighting should be community-driven, although I assume @hishamhm probably would not want to maintain syntax files for editors they do not use.

Some common editors:
+ Vim
+ Emacs
+ VSCode
+ Atom
+ Others?

Depends on #59.

**Comment by mrtnpwn** [comment]
I've just made one for GNU Nano, based off the original Lua syntax for Nano

https://github.com/mrtnpwn/tl-nano

**Comment by ildar** [comment]
+1 for [ZeroBraneStudio](https://studio.zerobrane.com/): awesome IDE!

**Comment by pdesaulniers** [comment]
I'll take care of VSCode support.

**Comment by pdesaulniers** [comment]
I've made a very basic [VSCode extension](https://github.com/pdesaulniers/vscode-tl). So far, it supports standard Lua syntax highlighting + a bare-bones language server that simply calls `tl check <buffer>`.

![2020-03-14-14:19:58](https://user-images.githubusercontent.com/23723605/76688020-790a3780-65ff-11ea-9b10-2ce2957b9816.png)

**Comment by hishamhm** [comment]
@pdesaulniers This is awesome!! 

Do you need the compiler to return the width of the error token, so that the red squiggle underlines the full token, or do you think this can be done in the extension itself?

**Comment by pdesaulniers** [comment]
> or do you think this can be done in the extension itself?

I ended up highlighting the "word" located at the error instead of just a single character. This gives much better results.

![2020-03-15-09:45:11](https://user-images.githubusercontent.com/23723605/76702665-e66ea480-66a1-11ea-83e9-8a25032ada17.png)

Today, I'll try to make the syntax highlighting work properly.

**Comment by pdesaulniers** [comment]
Mmmh, I wonder if we could use TL's lexer directly to provide syntax highlighting in VSCode... It seems like there is a 'semantic highlighting' API that we could use for this purpose, but it's fairly new: https://github.com/microsoft/vscode/wiki/Semantic-Highlighting-Overview

This would be much more simpler and maintainable than a TextMate grammar. I'll need to do some research.

**Comment by pdesaulniers** [comment]
The VS Code extension is now available in the marketplace: https://marketplace.visualstudio.com/items?itemName=pdesaulniers.vscode-teal

The syntax highlighting is still a work in progress, but it works for the most part.

**Comment by hishamhm** [comment]
@pdesaulniers Hooray!!

> @hishamhm, should I transfer the vscode-teal repo to the teal-language organization?

Yes, if you'd like I'd be happy to! I'll send you an invite to the org!

**Comment by resolritter** [comment]
Some thoughts about this

I first wondered if it would be difficult to adapt the [popular Lua syntax](https://github.com/tbastos/vim-lua/blob/master/syntax/lua.vim) for Vim, since Teal doesn't add a lot on top of it and Lua's syntax is, in general, pretty simple.

However, I've run into inconsistencies using [yats.vim](https://github.com/HerringtonDarkholme/yats.vim) (which is enormous, by the way, and [performance is an issue](https://github.com/HerringtonDarkholme/yats.vim/issues/96#issuecomment-466987828) due to how heavy it is). Granted, a lot of what makes that plugin heavy seems to be that it also supports JSX, which wouldn't be the case for Teal.

Enough to state, navigating around those heavy regex-based syntax engines is difficult ([see this question, for instance](https://stackoverflow.com/questions/27682324/set-item-to-higher-highlight-priority-on-vim/27687580), or [this plugin's interactions](https://github.com/luochen1990/rainbow/issues?q=is%3Aissue+is%3Aopen+label%3A%22confliction+with+other+third-party+syntax+plugins%22)) and requires lots of editor-specific knowledge to maintain. In summary, I fear that investing into a good syntax highlighting using regex-based definitions would run into the same problems and limitations I've described before; on the other hand, their editor support is noticeably widespread.

---

The good news: there's a MUCH better (and even easier, weirdly) option nowadays in [tree-sitter](https://github.com/tree-sitter/tree-sitter) ([talk](https://www.youtube.com/watch?v=Jes3bD6P0To)).

[Here's how you define the JS grammar](https://github.com/tree-sitter/tree-sitter-javascript/blob/master/grammar.js)

[And TypeScript's grammar is an extension of it](https://github.com/tree-sitter/tree-sitter-typescript/blob/a80ef55580201d55c3a40bdd7fc86f8bfe0dfde3/common/define-grammar.js#L24)

So here's why I think it might be easier: [there's already a grammar definition for Lua!](https://github.com/Azganoth/tree-sitter-lua/blob/master/grammar.js). So one could do the same and simply extend from it to create Teal's.

The methodology I'd take is:

- Copy over TypeScript's grammar
- Adapt it for the Lua grammar
- Throw `tl.tl` at it and start fixing/removing things until it compiles.

I'm interested in doing this myself, but I currently lack the knowledge necessary to make this happen. I'll see if I can get something started in the next few days.

**Comment by 3uclidian** [comment]
I've created a Vim syntax file and have included it in my
[BetterLua.vim](https://github.com/3uclidian/BetterLua.vim) plugin.
I'd say it's mostly complete, but the indent file needs some work, and I should
add some screenshots.

For Neovim in particular (which uses Lua for its api, and is a bit more modern
with its, well, everything):

-  a treesitter api is in the works, so a Teal parser would be nice to 
have some easier to maintain syntax highlighting.
- a language server client is in the works so a language server that 
doesn't need a js/ts dependency would be nice

**Comment by hishamhm** [comment]
> I've created a Vim syntax file and have included it in my BetterLua.vim plugin.

That's cool! Thank you for the plug! :)

> a treesitter api is in the works, so a Teal parser would be nice to have some easier to maintain syntax highlighting.

I've heard wonderful things about treesitter — this could be nice thing to have for incremental highlighting.

> a language server client is in the works so a language server that doesn't need a js/ts dependency would be nice

There was a Google Summer of Code project for a [pure-Lua language server for Typed Lua](https://github.com/mascarenhas/lua-lsp) (Teal's academic predecessor). I wonder if some of that code could be reused, but I haven't looked closely at it as I'm focusing on the compiler, but LSP support would be great to have!

**Comment by pdesaulniers** [comment]
@3uclidian Cool! 

> a language server client is in the works so a language server that doesn't need a js/ts dependency would be nice

I've made a simple ALE linter for Teal, which can be found in this plugin: https://github.com/teal-language/vim-teal

It's not very sophisticated (it simply calls `tl check`), but it's better than nothing.

**Comment by euclidianAce** [comment]
I mentioned this in the gitter, but i feel its relevant here too. I've started work on a tree-sitter parser for teal [here](https://github.com/euclidianAce/tree-sitter-teal) for the type-declaration branch. It currently parses `tl.tl` with no errors, but the names/nodes could do with some refactoring

**Comment by tooolbox** [comment]
This works for Sublime: https://gist.github.com/tooolbox/dc4f85dcd06ff79e08ed75b25d12533e

Haven't had the bandwidth to make it available on Package Control.

**Comment by tooolbox** [comment]
> Haven't had the bandwidth to make it available on Package Control.

If anyone wants to make a Sublime-Teal repo, I will contribute my syntax highlighting and my linter.  I would do it myself but would not be able to steward reliably, so I'm hoping someone else can.

**Comment by hishamhm** [comment]
I think this can be closed since we have syntax highlighting on various editors now, as shown in https://github.com/teal-language/tl/#text-editor-support and https://github.com/teal-language/awesome-teal !
